File: ios/Packages/AppFeatures/Sources/GamificationFeature/GamificationModels.swift
import Foundation

// MARK: - Local event queue (docs/features/gamification.md)

/// A locally queued, not-yet-confirmed-synced activity event.
public struct QueuedActivityEvent: Codable, Equatable, Sendable {
    public let clientEventId: String
    public let eventType: String
    public let occurredAt: Date
    public let timezone: String
    public let metadata: [String: String]

    public init(
        clientEventId: String = UUID().uuidString,
        eventType: String,
        occurredAt: Date = Date(),
        timezone: String = TimeZone.current.identifier,
        metadata: [String: String] = [:]
    ) {
        self.clientEventId = clientEventId
        self.eventType = eventType
        self.occurredAt = occurredAt
        self.timezone = timezone
        self.metadata = metadata
    }
}

/// Persistence boundary for the queue (mirrors TasbeehHistoryStoring).
public protocol ActivityEventQueueStoring: Sendable {
    func load() -> [QueuedActivityEvent]
    func save(_ events: [QueuedActivityEvent])
}

public struct FileActivityEventQueueStore: ActivityEventQueueStoring {
    private let fileURL: URL

    public init(directory: URL) {
        fileURL = directory.appendingPathComponent("gamification-event-queue.json")
    }

    public func load() -> [QueuedActivityEvent] {
        guard let data = try? Data(contentsOf: fileURL) else { return [] }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return (try? decoder.decode([QueuedActivityEvent].self, from: data)) ?? []
    }

    public func save(_ events: [QueuedActivityEvent]) {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(events) else { return }
        try? FileManager.default.createDirectory(
            at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try? data.write(to: fileURL, options: .atomic)
    }
}

// MARK: - Server profile (GET /v1/gamification/profile)

public struct GamificationStreak: Decodable, Equatable, Identifiable, Sendable {
    public var id: String { key }
    public let key: String
    public let name: String
    public let currentLength: Int
    public let longestLength: Int
    public let graceRemaining: Int

    enum CodingKeys: String, CodingKey {
        case key, name
        case currentLength = "current_length"
        case longestLength = "longest_length"
        case graceRemaining = "grace_remaining"
    }
}

public struct GamificationMission: Decodable, Equatable, Identifiable, Sendable {
    public var id: String { key }
    public let key: String
    public let name: String
    public let progress: Int
    public let target: Int
    public let window: String
    public let endsAt: Date?

    enum CodingKeys: String, CodingKey {
        case key, name, progress, target, window
        case endsAt = "ends_at"
    }
}

public struct GamificationBadge: Decodable, Equatable, Identifiable, Sendable {
    public var id: String { key }
    public let key: String
    public let name: String
    public let iconRef: String
    public let earnedAt: Date?

    enum CodingKeys: String, CodingKey {
        case key, name
        case iconRef = "icon_ref"
        case earnedAt = "earned_at"
    }

    public var isEarned: Bool { earnedAt != nil }
}

public struct GamificationProfile: Decodable, Equatable, Sendable {
    public let streaks: [GamificationStreak]
    public let missions: [GamificationMission]
    public let badges: [GamificationBadge]

    public static let empty = GamificationProfile(streaks: [], missions: [], badges: [])
}

File: android/feature/gamification/src/main/kotlin/com/fatwabot/feature/gamification/GamificationModels.kt
package com.fatwabot.feature.gamification

import java.io.File
import java.util.UUID
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/** A locally queued, not-yet-confirmed-synced activity event. */
@Serializable
data class QueuedActivityEvent(
    val clientEventId: String = UUID.randomUUID().toString(),
    val eventType: String,
    val occurredAtEpochSeconds: Long,
    val timezone: String,
    val metadata: Map<String, String> = emptyMap(),
)

/** Persistence boundary for the queue (mirrors TasbeehHistoryStoring). */
interface ActivityEventQueueStoring {
    fun load(): List<QueuedActivityEvent>
    fun save(events: List<QueuedActivityEvent>)
}

class FileActivityEventQueueStore(private val file: File) : ActivityEventQueueStoring {
    private val json = Json { ignoreUnknownKeys = true }
    private val serializer = ListSerializer(QueuedActivityEvent.serializer())

    override fun load(): List<QueuedActivityEvent> = runCatching {
        if (!file.exists()) emptyList() else json.decodeFromString(serializer, file.readText())
    }.getOrDefault(emptyList())

    override fun save(events: List<QueuedActivityEvent>) {
        runCatching {
            file.parentFile?.mkdirs()
            val tmp = File(file.parentFile, "${file.name}.tmp")
            tmp.writeText(json.encodeToString(serializer, events))
            tmp.renameTo(file)
        }
    }
}

// Server profile (GET /v1/gamification/profile)

@Serializable
data class GamificationStreak(
    val key: String,
    val name: String,
    @SerialName("current_length") val currentLength: Int,
    @SerialName("longest_length") val longestLength: Int,
    @SerialName("grace_remaining") val graceRemaining: Int,
)

@Serializable
data class GamificationMission(
    val key: String,
    val name: String,
    val progress: Int,
    val target: Int,
    val window: String,
    @SerialName("ends_at") val endsAt: String? = null,
)

@Serializable
data class GamificationBadge(
    val key: String,
    val name: String,
    @SerialName("icon_ref") val iconRef: String,
    @SerialName("earned_at") val earnedAt: String? = null,
) {
    val isEarned: Boolean get() = earnedAt != null
}

@Serializable
data class GamificationProfile(
    val streaks: List<GamificationStreak> = emptyList(),
    val missions: List<GamificationMission> = emptyList(),
    val badges: List<GamificationBadge> = emptyList(),
) {
    companion object {
        val EMPTY = GamificationProfile()
    }
}
